book/src/dev/rfcs/0005-state-updates.md

```diff
@@ -524,49 +524,62 @@ The state service uses the following entry points:
 
 ## Committing non-finalized blocks
 
-Given the above structures for manipulating the non-finalized state new
-`non-finalized` blocks are commited as follows:
+New `non-finalized` blocks are commited as follows:
 
-### `fn queue_and_commit_non_finalized_blocks(&mut self, new: QueuedBlock)`
+### `pub(super) fn queue_and_commit_non_finalized_blocks(&mut self, new: Arc<Block>) -> tokio::sync::broadcast::Receiver<block::Hash>`
 
-1. If the block itself exists in the finalized chain, it has already been
-   successfully verified:
-  - broadcast `Ok(block.hash())` via `block.rsp_tx`, and return
+1. If a duplicate block exists in the queue:
+     - Find the `QueuedBlock` for that existing duplicate block
+     - Create an extra receiver for the existing block, using `block.rsp_tx.subscribe`,
+     - Drop the newly received duplicate block
+     - Return the extra receiver, so it can be used in the response future for the duplicate block request
 
-2. Add `block` to `self.queued_blocks`
+2. Create a `QueuedBlock` for `block`:
+     - Create a `tokio::sync::broadcast` channel
+     - Use that channel to create a `QueuedBlock` for `block`.
 
-3. If `block.header.previous_block_hash` is not present in the finalized or
-   non-finalized state return early
+3. If a duplicate block exists in a non-finalized chain, or the finalized chain,
+   it has already been successfully verified:
+     - Broadcast `Ok(block.hash())` via `block.rsp_tx`, and return the receiver for the block's channel
 
-4. Else iteratively attempt to process queued blocks by their parent hash
+4. Add `block` to `self.queued_blocks`
```

Comment:
We can split this out into a new function `queue_block`, which:
- takes a `QueuedBlock`
- if the previous block is found, returns the `previous_block_hash` for further processing

(We don't have to make this change in the spec, it's enough to just make it in the code.)